### PR TITLE
Collision solver

### PR DESCRIPTION
**Pull request by Thomyrock**
This is an attempt to code what happens when Isaac, or another entity, collides with an element

### CODE REVIEW

File: doc/CollisionSolverEntities.doc.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+}
+~~~
+
+## Note
+
+Same function as for Isaac, but with less cases here (other entities donc take contact damages, and can't activate anything
+
+## TODO
+
+* the rollback technic to manage collision with blocking elements is likely not to be the best, as it can cause lines of uncrossable pixels between Isaac and the element.
```
Comment (DorianXGH) [inline doc/CollisionSolverEntities.doc.md:43]: the rollback techniQUE

File: doc/CollisionSolverIsaac.doc.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}
+~~~
+
+## Note
+
+We mustn't forget to decrease the recover counter at each VBLank call
+
+## TODO
+
+* the rollback technic to manage collision with blocking elements is likely not to be the best, as it can cause lines of uncrossable pixels between Isaac and the element.
```
Comment (DorianXGH) [inline doc/CollisionSolverIsaac.doc.md:49]: same

File: src/CollisionSolverEntities.s [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	collidingEntity DW
+.ENDE
+
+; //// Collision Solver for Entities other than Isaac \\\\
+collisionSolverIsaac :
+	push c
+	push d
+	push e
+	
+	ld c, hl ;get element's size and flags
```
Comment (DorianXGH) [inline src/CollisionSolverEntities.s:11]: error : 16 bits loading into a 8bit register : does not fit

File: src/CollisionSolverEntities.s [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+	;// Touch Horizontally \\
+	bit 1, b
+	jr z, noEHorizontalCollision
+	ld a, (hl)
+	and %11110000
+	srl a
+	srl a
+	srl a
+	srl a
```
Comment (DorianXGH) [inline src/CollisionSolverEntities.s:28]: you can use SWAP in this case (4 bitshifts to the right : SWAP the nibbles and AND with 0x0F)

File: src/CollisionSolverIsaac.s [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+	;// Touch Horizontally \\
+	bit 1, b
+	jr z, noHorizontalCollision
+	ld a, (isaac.speed)
+	and %11110000
+	srl a
+	srl a
+	srl a
+	srl a
```
Comment (DorianXGH) [inline src/CollisionSolverIsaac.s:23]: same

File: src/CollisionSolverIsaac.s [hunk 6]
```diff
@@ -0,0 +1,82 @@
+.DEFINE RECOVERYTIME 30
+
+; //// Collision Solver for Isaac \\\\
+collisionSolverIsaac :
+	push c
+	push d
+	push e
+	
+	ld c, hl ;get element's size and flags
```
Comment (DorianXGH) [inline src/CollisionSolverIsaac.s:9]: Error here too

File: doc/CollisionSolverEntities.doc.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+None
+
+## Pseudo Code
+
+~~~C
+// a function for when an entity collides with any element
+char void solve_collision_to_Isaac (
+	element entity,
+	char touched_from,
+	element.sheet* s) {
```
Comment (JeanMarieMineau) [inline doc/CollisionSolverEntities.doc.md:26]: Make more clear the link between the function parameters (like entity) and those given in the Parameter section.

File: doc/CollisionSolverEntities.doc.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
+# Collision Solver for Entities other than Isaac
+
+## Label
+
+`collision`
```
Comment (JeanMarieMineau) [inline doc/CollisionSolverEntities.doc.md:5]: The label is the label used to call the function, here I think it's `solve_collision_to_Isaac`?
Anyway `collusion` is used by the collision function.

File: doc/CollisionSolverEntities.doc.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+## Label
+
+`collision`
+
+## Parameters
+
+| Label | Type | Size/Struct | Description |
+| ----- | ---- | ----------- | ----------- |
+| b | register | 1 byte | the 4 left bits are flags which indicates which side of the element the entity touched (right, left, lower, upper) |
```
Comment (JeanMarieMineau) [inline doc/CollisionSolverEntities.doc.md:11]: I don't think we know how to find this information yet, can you find a solution to make it works without?

File: doc/CollisionSolverIsaac.doc.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+## Label
+
+`collision`
+
+## Parameters
+
+| Label | Type | Size/Struct | Description |
+| ----- | ---- | ----------- | ----------- |
+| b | register | 1 byte | the 2 left bits tell whether the collision happens during (respectively) horizontal or vertical movement |
```
Comment (JeanMarieMineau) [inline doc/CollisionSolverIsaac.doc.md:11]: Same things here, I'm afraid we can't have this information with the current collision function.

File: doc/CollisionSolverIsaac.doc.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+None
+
+## Pseudo Code
+
+~~~C
+// a function for when Isaac collides with any element
+char void solve_collision_to_Isaac (
+	char touched_from,
+	element.sheet* s) {
```
Comment (JeanMarieMineau) [inline doc/CollisionSolverIsaac.doc.md:24]: Again, explicit the link between b, hl, touched_from and s please

File: src/CollisionSolverEntities.lib.s [hunk 12]
```diff
@@ -0,0 +1,48 @@
+; //// Collision Solver for Entities other than Isaac \\\\
+collisionSolverIsaac :
+	push c
+	push d
+	push e
```
Comment (JeanMarieMineau) [inline src/CollisionSolverEntities.lib.s:5]: e is not used, no need to save it

File: src/CollisionSolverEntities.lib.s [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+	dec hl ;hl leads to  entity's y coordinate
+	ld a, (hl)
+	sub d
+	ld (hl), a
+	; \\ Touch Vertically //
+@noEVerticalCollision :
+	;\\\ Blocking Collision ///
+@noEBlockCollision :
+
+	pop e
```
Comment (JeanMarieMineau) [inline src/CollisionSolverEntities.lib.s:44]: e is not used, no need to save it

File: src/CollisionSolverIsaac.lib.s [hunk 14]
```diff
@@ -0,0 +1,79 @@
+.DEFINE RECOVERYTIME 30
+
+; //// Collision Solver for Isaac \\\\
+collisionSolverIsaac :
+	push c
+	push d
+	push e
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:7]: same here

File: src/CollisionSolverEntities.lib.s [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+	push e
+	
+	ld c, (hl) ;get element's size and flags
+	ld hl, (collidingEntity)
+	inc hl
+	inc hl ;hl leads to entity's speed
+	
+	;/// Blocking Collision \\\
+	bit 4, c
+	jp z, noEBlockCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverEntities.lib.s:14]: @noEBlockCollision

File: src/CollisionSolverEntities.lib.s [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+	;// Touch Horizontally \\
+	bit 7, b
+	jr z, noEHorizontalCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverEntities.lib.s:18]: @noEHorizontalCollision

File: src/CollisionSolverEntities.lib.s [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+	ld d, a
+	ld a, (collidingEntity)
+	sub d
+	ld (collidingEntity), a
+	;\\ Touch Horizontally //
+@noEHorizontalCollision :
+
+	; // Touch Vertically \\
+	bit 6, b
+	jr z, noEVerticalCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverEntities.lib.s:31]: @noEVerticalCollision

File: src/CollisionSolverIsaac.lib.s [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+collisionSolverIsaac :
+	push c
+	push d
+	push e
+	
+	ld c, (hl) ;get element's size and flags
+	
+	;/// Blocking Collision \\\
+	bit 4, c
+	jp z, noBlockCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:13]: @noBlockCollision

File: src/CollisionSolverIsaac.lib.s [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+	;// Touch Horizontally \\
+	bit 7, b
+	jr z, noHorizontalCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:17]: @noHorizontalCollision

File: src/CollisionSolverIsaac.lib.s [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+	ld d, a
+	ld a, (isaac.x)
+	sub d
+	ld (isaac.x), a
+	;\\ Touch Horizontally //
+@noHorizontalCollision :
+
+	; // Touch Vertically \\
+	bit 6, b
+	jr z, noVerticalCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:30]: @noVerticalCollision

File: src/CollisionSolverIsaac.lib.s [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+	; \\ Touch Vertically //
+@noVerticalCollision :
+	;\\\ Blocking Collision ///
+@noBlockCollision :
+	
+	ld de, hl ;saving sheet address
+	
+	;/// Hurting Collision \\\
+	bit 3, c
+	jr z, noHurtCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:46]: @noHurtCollision

File: src/CollisionSolverIsaac.lib.s [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+	inc hl
+	ld a, (isaac.hp)
+	sub (hl)
+	ld (isaac.hp), a
+	
+	;starting recovery
+	ld a, RECOVERYTIME
+	ld (isaac.recover), a
+	;\\\ Hurting Collision ///
+noHurtCollision :
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:61]: @noHurtCollision

File: src/CollisionSolverIsaac.lib.s [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+@noBlockCollision :
+	
+	ld de, hl ;saving sheet address
+	
+	;/// Hurting Collision \\\
+	bit 3, c
+	jr z, noHurtCollision
+	ld a, (isaac.recover)
+	cp 0
+	jr z, noHurtCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:49]: @noHurtCollision

File: src/CollisionSolverIsaac.lib.s [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+	
+	;starting recovery
+	ld a, RECOVERYTIME
+	ld (isaac.recover), a
+	;\\\ Hurting Collision ///
+noHurtCollision :
+	
+	; /// Activation Collision \\\
+	bit 2, c
+	jr z, noReactCollision
```
Comment (JeanMarieMineau) [inline src/CollisionSolverIsaac.lib.s:65]: @noReactCollision

File: doc/CollisionSolverEntities.doc.md [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+| hl | register | 2 bytes | address to the caracter's sheet of the element |
+| collisionSolver_.collidingEntity | fixed address | element pointer | address to the element representing the entity |
+
+## Return
+
+None
+
```
Comment (JeanMarieMineau) [inline doc/CollisionSolverEntities.doc.md:18]: Add a link or a copy of the used structs.

File: src/CollisionSolverIsaac.lib.s (hunk 14 above)
Comment (Thomyrock) [reply src/CollisionSolverIsaac.lib.s:7]: In this function e is overwritten (line 42)

**Pull request merged**